exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+            input_size=n_dims,
+            hidden_size=hidden_size,
+            num_layers=num_layers,
+            bias=bool(bias),
+            batch_first=True,
+            dropout=dropout,
+            bidirectional=bool(bidirectional),
+        )
+        n_rnn_out_features = hidden_size * (2 if bidirectional else 1)
+        self.decoder = nn.Linear(in_features=n_rnn_out_features, out_features=n_tokens)
+        self.log_softmax = nn.LogSoftmax(dim=2)
+
+    def forward(self, input, hidden_states=None, return_hidden=False):
+        batch_size = input.size(0 if self.rnn.batch_first else 1)
+        n_directions = 2 if self.rnn.bidirectional else 1
+        full_num_layers = self.rnn.num_layers * n_directions
+        if hidden_states is None:
+            hidden_states = torch.zeros(
+                full_num_layers, batch_size, self.rnn.hidden_size
+            ).to(input.device, non_blocking=True)
+        output = input
+        output = self.encoder(output)
+        output, new_hidden_states = self.rnn(
+            output,
+            hidden_states.view(full_num_layers, batch_size, self.rnn.hidden_size),
+        )
+        output = self.decoder(output)
+        output = self.log_softmax(output)
+        if return_hidden:
+            new_hidden_states = new_hidden_states.view(hidden_states.size())
+            return output, new_hidden_states
+        else:
+            return output
+
+
+class LossFunction(nn.Module):
+    """
+    Module to compute the aggregate (over all elements of a batch, and all
+    sequence positions within each element) loss.
+    """
+
+    def __init__(self, reduction="mean"):
+        super().__init__()
+        self.loss_fn = nn.NLLLoss(reduction=reduction)
+
+    def forward(self, input, target):
+        return self.loss_fn(input.view(target.numel(), -1), target.view(-1))
+
+
+class ModelManager:
+    """
+    Objects that own torch models (modules) and provide convenience methods
+    for their training and use.
+    """
+
+    def __init__(
+        self,
+        model=None,
+        device=torch.device("cuda:0"),
+        **kwargs
+    ):
+        if model is None:
+            self.model = LanguageModel(**kwargs)
+        else:
+            self.model = model
+        self.device = device
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, new_device):
+        if isinstance(new_device, str):
+            new_device = torch.device(new_device)
+        if new_device != torch.device("cpu") and not torch.cuda.is_available():
+            print("Warning:  cuda not available so setting device to CPU.")
+            self._device = torch.device("cpu")
+        else:
+            self._device = new_device
+        self.model.to(self._device)
+
+    def collate_batch_with_labels(self, raw_batch):
+        """
+        Take a raw batch (as returned by a DataLoader and SpmIdsBatchSampler),
+        collate it, and split off labels.
```

Comment:
It could be clearer that "labels" here are the ID of the next token after this position.

Reply:
(After all, there was some confusion on this when we were originally discussing how the language model should work.)

Reply:
I've refactored the code so that this function is now a member of another class (and renamed it); I hope that makes what is going on more clear.  (But let me know if you still find room for improvement.)